### PR TITLE
add stop/restart support;add archival config flag;

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1521,7 +1521,6 @@ func (bc *BlockChain) InsertHeaderChain(chain []*types.Header, checkFreq int) (i
 	if i, err := bc.hc.ValidateHeaderChain(chain, checkFreq); err != nil {
 		return i, err
 	}
-
```
Comment (LeoHChen) [inline core/blockchain.go]: ?

File: node/node.go [hunk 2]
```diff
@@ -228,16 +233,21 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, db ethdb.Database) *N
 	if host != nil && consensusObj != nil {
 		// Consensus and associated channel to communicate blocks
 		node.Consensus = consensusObj
-
-		// Init db.
```
Comment (LeoHChen) [inline node/node.go]: ?

File: node/node.go [hunk 3]
```diff
@@ -228,16 +233,21 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, db ethdb.Database) *N
...
 		database := db
 		if database == nil {
 			database = ethdb.NewMemDatabase()
+			chain, err = node.GenesisBlockSetup(database)
+			isFirstTime = true
+		} else {
+			chain, err = node.InitBlockChainFromDB(db, node.Consensus)
+			isFirstTime = false
+			if err != nil || chain == nil || chain.CurrentBlock().NumberU64() <= 0 {
+				chain, err = node.GenesisBlockSetup(database)
```
Comment (LeoHChen) [inline node/node.go]: do you need a log line so that we know what's happening?

File: node/node.go [hunk 4]
```diff
@@ -461,3 +472,13 @@ func (node *Node) AddBeaconChainDatabase(db ethdb.Database) {
 	node.beaconChain = chain
 	node.BeaconWorker = worker.New(params.TestChainConfig, chain, node.Consensus, pki.GetAddressFromPublicKey(node.SelfPeer.ConsensusPubKey), node.Consensus.ShardID)
 }
+
+func (node *Node) InitBlockChainFromDB(db ethdb.Database, consensus *consensus.Consensus) (*core.BlockChain, error) {
```
Comment (LeoHChen) [inline node/node.go]: you need comment on public function.

File: node/node.go [hunk 5]
```diff
@@ -461,3 +472,13 @@ func (node *Node) AddBeaconChainDatabase(db ethdb.Database) {
...
+	chainConfig := params.TestChainConfig
+	if consensus != nil {
+		chainConfig.ChainID = big.NewInt(int64(consensus.ShardID)) // Use ChainID as piggybacked ShardID
+	}
+	cacheConfig := core.CacheConfig{Disabled: false, TrieNodeLimit: 256 * 1024 * 1024, TrieTimeLimit: 5 * time.Minute}
```
Comment (LeoHChen) [inline node/node.go]: please use some constants here.

File: core/blockchain.go (hunk 1 above)
Comment (chaosma) [reply core/blockchain.go]: I made several debug info in the code, not cleaned up fully

File: node/node.go (hunk 4 above)
Comment (chaosma) [reply node/node.go]: added because of the travis check

File: node/node.go (hunk 3 above)
Comment (chaosma) [reply node/node.go]: If the local database is available, we will create the genesis block. it's not an error

File: node/node.go (hunk 5 above)
Comment (chaosma) [reply node/node.go]: This is the default value from Ethereum code. They didn't use constant either. This is a placeholder. After we add the archival node role, I will remove it and add config for archival node.

File: core/blockchain.go (hunk 1 above)
Comment (chaosma) [reply core/blockchain.go]: add space back

File: node/node.go (hunk 2 above)
Comment (chaosma) [reply node/node.go]: add back

File: node/node.go (hunk 3 above)
Comment (LeoHChen) [reply node/node.go]: Yes, I understand.  Please just add an informational log.

**Comment by chaosma** [comment]
Yes, it will read the backup file by default.

**Pull request merged**